### ISSUE TITLE
Add generic overloads for Delegate.CreateDelegate

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/corefx#40941]
Currently usages of `Delegate.CreateDelegate` tend to be along these lines:

`(Func<object>)Delegate.CreateDelegate(typeof(FuncObject), methodInfo)`

For complicated delegates, this can be quite annoying and error prone.

it would be extremely simple to add generic overloads

```csharp
T CreateDelegate<T>(MethodInfo methodInfo) where T : Delegate
{
    return (T)CreateDelegate(typeof(T), methodInfo);
}
```
and it would make the API much nicer to use.

#### API suggestion

The API is simple. We provide a matching generic overload for every existing overload of Delegate.CreateDelegate, which takes a generic type parameter instead of its first parameter, and returns an instance of the type parameter.

 The implementation for them all is trivial.

**Comment by DaZombieKiller** [comment]
+1, I often find myself writing extension methods on `MethodInfo` for this purpose.

**Comment by GrabYourPitchforks** [comment]
How can we make progress on this? The API proposal is straightforward enough and normally it'd be actionable by us. But it sounds like there's concern that the proposal as currently stated is incomplete?

**Comment by jkotas** [comment]
@GrabYourPitchforks The proposal says to add generic overload to every CreateDelegate variant:

- It is not possible to add it everywhere because of it creates ambiguity. We need to figure how to deal with this ambiguity.
- It probably does not make sense to add it to every existing CreateDelegate variant. There is a lot of redundancy in the existing CreateDelegate variants. My proposal above was to add it for `MethodInfo.CreateDelegate` only.

**Comment by bartonjs** [comment]
Can you expand on the ambiguity, Jan?

Given my very clever class of:

```C#
    public class Potato
    {
        public object Cook(Type t) => t;
        public object Cook(Type t, object o) => Tuple.Create(t, o);
        public T Cook<T>() => default;
        public T Cook<T>(object o) => default;
    }
```

I don't have compile-time or runtime ambiguity with these calls (intentionally using var to maximize ambiguity):

```C#
Potato p = new Potato();
var one = p.Cook<Action>();
var two = p.Cook<Action>(one);
var three = p.Cook(typeof(Action));
var four = p.Cook(typeof(Action), three);
var five = p.Cook(null);
```

Since the T isn't used in a parameter it isn't a candidate for generic inference, so there's no inference-based ambiguity (and even if there was, the non-generic method wins).

But if I'm confused and you're already suggesting that we do

```C#
public partial class MethodInfo
{
    // Current
    public virtual Delegate CreateDelegate(Type delegateType);
    public virtual Delegate CreateDelegate(Type delegateType, object target);

    // New
    public T CreateDelegate<T>() => (T)CreateDelegate(typeof(T));
    public T CreateDelegate<T>(object? target) => (T)CreateDelegate(typeof(T), target);
}
```

Then I'm on board.

Unless it's ambiguous in F# or VB?

**Comment by jkotas** [comment]
@bartonjs You are right. I am not sure what I was thinking about. So two new generic methods on MethodInfo is the plan then?

**Comment by bartonjs** [comment]
Added generic constraints (where T : Delegate) and updated the top post.

**Comment by GrabYourPitchforks** [comment]
Any issues with adding generic instance methods to a non-generic type? I recall some discussion in the past about trying to avoid such things.

**Comment by jkotas** [comment]
Virtual generic methods are problematic. Non-virtual generic instance methods are fine.
